### PR TITLE
[SPARK-49902][SQL] Catch underlying runtime errors in `RegExpReplace`

### PR DESCRIPTION
**Pull request by harshmotw-db**
### What changes were proposed in this pull request?

Earlier, runtime errors in underlying libraries were not caught during runtime in the RegExpReplace expression. The underlying errors were thrown directly to the user. For example, it wouldn't be uncommon to see issues like `java.lang.IndexOutOfBoundsException: No group 3`. This PR introduces a change to catch these underlying issues and throw a SparkException instead which details the input on which the exception failed. The new Spark Exception looks something like `org.apache.spark.SparkException: Could not perform regexp_replace for source = <source>, pattern = 
<pattern>, replacement = <replacement> and position = <position>`.

### Why are the changes needed?

Two reasons. First, the new exception details which row the given error occurred on, which makes it easier for the user to debug the query or Spark developers to identify bugs. Second, a Spark Exception is generally considered expected behavior indicating that there were no unintended issues in the query's execution.

### Does this PR introduce _any_ user-facing change?

Yes, a better exception is thrown when RegExpReplace fails.

### How was this patch tested?

Unit test in both codegen as well as interpreted mode.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by harshmotw-db** [comment]
@cloud-fan Can you look at this PR? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 1]
```diff
@@ -700,7 +701,14 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
       m.region(position, source.length)
       result.delete(0, result.length())
       while (m.find) {
-        m.appendReplacement(result, lastReplacement)
+        try {
+          m.appendReplacement(result, lastReplacement)
+        } catch {
+          case e: Exception =>
+            throw new SparkException(s"Could not perform regexp_replace for " +
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:708]: Please, introduce an error condition (class) for the error.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 1 above)
Comment (harshmotw-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Thanks for the suggestion, I have incorporated it

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala [hunk 2]
```diff
@@ -96,6 +98,31 @@ class CollationSQLRegexpSuite
...
+          val df = sql(query)
+          val plan = df.queryExecution.executedPlan
+          assert(plan.isInstanceOf[WholeStageCodegenExec] == (codegenMode == "CODEGEN_ONLY"))
+          val exception = intercept[SparkRuntimeException] {
+            df.collect()
+          }
+          assert(exception.getCondition == "INVALID_REGEXP_REPLACE")
+          assert(exception.getMessage.contains("""Could not perform regexp_replace for """ +
+            """`source = "first last"`, `pattern = "(?<first>[a-zA-Z]+) (?<last>[a-zA-Z]+)"`, """ +
+            """`replacement = "$3 $1"` and `position = 1`."""))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:119]: Could you use `checkError`, please. We try to avoid embedding error messages to source code because of:
- tech editors and others should be able to edit error formats without repackaging Spark
- we might have error messages in other languages in the future.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 3]
```diff
@@ -2907,6 +2907,12 @@
     ],
     "sqlState" : "42613"
   },
+  "INVALID_REGEXP_REPLACE" : {
+    "message" : [
+      "Could not perform regexp_replace for `source = \"<source>\"`, `pattern = \"<pattern>\"`, `replacement = \"<replacement>\"` and `position = <position>`."
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-conditions.json:2912]: Please, remove backticks. They are not necessary here, and we use them to highlight message parameters, literals and so on.

Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-conditions.json:2912]:
```suggestion
      "Could not perform `regexp_replace` for source = \"<source>\", pattern = \"<pattern>\", replacement = \"<replacement>\" and position = <position>."
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 4]
```diff
@@ -700,7 +700,13 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
       m.region(position, source.length)
       result.delete(0, result.length())
       while (m.find) {
-        m.appendReplacement(result, lastReplacement)
+        try {
+          m.appendReplacement(result, lastReplacement)
+        } catch {
+          case e: Exception =>
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:706]: can't you catch more specific exception here?

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala [hunk 5]
```diff
@@ -96,6 +98,31 @@ class CollationSQLRegexpSuite
...
 
+  test("RegExpReplace throws the right exception when replace fails on a particular row") {
+    val tableName = "regexpReplaceException"
+    withTable(tableName) {
+      Seq("NO_CODEGEN", "CODEGEN_ONLY").foreach { codegenMode =>
+        withSQLConf("spark.sql.codegen.factoryMode" -> codegenMode) {
+          sql(s"CREATE TABLE IF NOT EXISTS $tableName(s STRING)")
+          sql(s"INSERT INTO $tableName VALUES('first last')")
+          val query = s"SELECT regexp_replace(s, '(?<first>[a-zA-Z]+) (?<last>[a-zA-Z]+)', " +
+            s"'$$3 $$1') FROM $tableName"
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:109]: I don't know. Is there any other databases support `regexp_replace(s, '(?<first>[a-zA-Z]+) (?<last>[a-zA-Z]+)', '$$3 $$1')` ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 4 above)
Comment (harshmotw-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:706]: I'm not 100% sure. Based on what I can see in the underlying Java file, I think the most restrictive choice we can make is `RuntimeException` which is just one level above `Exception`. But I still think `Exception` is the safest choice since there might be potential errors that don't fall in the `RuntimeExceptions` category.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala (hunk 5 above)
Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:109]: I guess the designer of `RegExpReplace` didn't realize the feature(In fact, a bug here). We should forbid it or consider a better way.

Comment (beliefer) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:109]: Personally, I think check the rep expression if contains $1, $2 ... is a easier method.

Comment (harshmotw-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:109]: @beliefer The designer of `RegExpReplace` very well may have thought of this case and in my opinion it works exactly as intended. It's just that sometimes these patterns are out of index and we just throw the library error in this case. This PR aims to handle these exceptions better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 4 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:706]: at least catch non-fatal exceptions like:
```suggestion
          case NonFatal(e) =>
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala [hunk 6]
```diff
@@ -96,6 +98,36 @@ class CollationSQLRegexpSuite
     }
   }
 
+  test("RegExpReplace throws the right exception when replace fails on a particular row") {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:101]: Why did you place this test to collation related test suite, I wonder.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala [hunk 7]
```diff
@@ -96,6 +98,36 @@ class CollationSQLRegexpSuite
...
+    val tableName = "regexpReplaceException"
+    withTable(tableName) {
+      Seq("NO_CODEGEN", "CODEGEN_ONLY").foreach { codegenMode =>
+        withSQLConf("spark.sql.codegen.factoryMode" -> codegenMode) {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:105]:
```suggestion
        withSQLConf(SQLConf.CODEGEN_FACTORY_MODE.key -> codegenMode) {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala [hunk 8]
```diff
@@ -96,6 +98,36 @@ class CollationSQLRegexpSuite
...
+          sql(s"CREATE TABLE IF NOT EXISTS $tableName(s STRING)")
+          sql(s"INSERT INTO $tableName VALUES('first last')")
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:107]: You don't need to create the table in the loop. Please, move to the upper level.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala (hunk 6 above)
Comment (harshmotw-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:101]: Good question. I didn't want to create a whole new test suite for this small test. Therefore, it put it in the only SQL test suite I could find for regexp-expressions. The other `RegexpExpressionsSuite.scala` doesn't do end-to-end SQL testing.

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:101]: How about https://github.com/apache/spark/blob/a1fc7e66883ac61abfd4be15f50455d03cac040c/sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala#L180

Comment (harshmotw-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala]: Thanks, done!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 9]
```diff
@@ -748,7 +755,16 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
         $matcher.region($position, $source.length());
 
         while ($matcher.find()) {
-          $matcher.appendReplacement($termResult, $termLastReplacement);
+          try {
+            $matcher.appendReplacement($termResult, $termLastReplacement);
+          } catch (Exception e) {
+            if (scala.util.control.NonFatal.apply(e)) {
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:761]: Please use `catch (Throwable e)` here.

Comment (harshmotw-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:761]: You're right. Resolved.

**Comment by harshmotw-db** [comment]
@beliefer @MaxGekk I don't believe the linters fail has anything to do with this PR. The error says:
```
python/pyspark/sql/types.py:3276: error: Module has no attribute "isdtype"  [attr-defined]
```
and the PR doesn't make any changes in types.py.

**Comment by MaxGekk** [comment]
@harshmotw-db Could you re-trigger only the failed GitHub action, please.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala [hunk 10]
```diff
@@ -19,7 +19,7 @@ package org.apache.spark.sql
 
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.plans.logical.Project
-import org.apache.spark.sql.internal.SqlApiConf
+import org.apache.spark.sql.internal.{SqlApiConf}
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:22]: nit: unnecessary changes.

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLRegexpSuite.scala:22]:
```suggestion
import org.apache.spark.sql.internal.SqlApiConf
```

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @harshmotw-db and @beliefer for review.
